File: client/src/main/java/net/treset/minecraftlauncher/sync/ManifestSynchronizer.java
package net.treset.minecraftlauncher.sync;

import javafx.util.Pair;
import net.treset.mc_version_loader.json.SerializationException;
import net.treset.mc_version_loader.launcher.LauncherManifest;
import net.treset.mc_version_loader.launcher.LauncherManifestType;
import net.treset.mc_version_loader.util.DownloadStatus;
import net.treset.minecraftlauncher.LauncherApplication;
import net.treset.minecraftlauncher.data.LauncherFiles;
import net.treset.minecraftlauncher.util.file.LauncherFile;
import net.treset.minecraftlauncher.util.string.FormatString;
import net.treset.minecraftlauncher.util.string.UrlString;

import java.io.File;
import java.io.IOException;
import java.nio.file.Files;
import java.util.ArrayList;
import java.util.List;

public class ManifestSynchronizer extends FileSynchronizer {
    private LauncherManifest manifest;
    protected LauncherFiles files;

    public ManifestSynchronizer(LauncherManifest manifest, LauncherFiles files, SyncCallback callback) {
        super(callback);
        this.manifest = manifest;
        this.files = files;
    }

    @Override
    public void upload() throws IOException {
        if(SyncService.isSyncing(manifest)) {
            uploadExisting();
        } else {
            uploadNew();
        }

    }

    @Override
    public void download() throws IOException {
        if(SyncService.isSyncing(manifest)) {
            downloadExisting();
        } else {
            downloadNew();
        }
    }

    protected void uploadExisting() throws IOException {
        setStatus(new SyncStatus(SyncStep.COLLECTING, null));
        ComponentData currentData = getCurrentComponentData();
        ComponentData newData = calculateComponentData(currentData.getVersion());
        List<String> difference = compareHashes(new ComponentData.HashEntry("", currentData.getHashTree()), new ComponentData.HashEntry("", newData.getHashTree()), LauncherFile.of("./"));
        uploadDifference(difference);
        completeUpload(newData);
        setStatus(new SyncStatus(SyncStep.FINISHED, null));
    }

    protected void uploadNew() throws IOException {
        setStatus(new SyncStatus(SyncStep.STARTING, null));
        LOGGER.debug("Sync file not found, creating new sync file");
        setStatus(new SyncStatus(SyncStep.COLLECTING, null));
        ComponentData data = calculateComponentData(0);
        setStatus(new SyncStatus(SyncStep.UPLOADING, null));
        SyncService service = new SyncService();
        service.newComponent(SyncService.convertType(manifest.getType()), manifest.getId());
        uploadAll(data);
        LOGGER.debug("Completing upload...");
        int version = completeUpload(data);
        LOGGER.debug("Upload complete: version=" + version);
        setStatus(new SyncStatus(SyncStep.FINISHED, null));
    }

    protected void downloadExisting() throws IOException {
        ComponentData data = getCurrentComponentData();

        downloadDifference(data.getVersion());
    }

    protected void downloadNew() throws IOException {
        File dir = new File(manifest.getDirectory());
        if(!dir.exists()) {
            Files.createDirectories(dir.toPath());
        }

        downloadDifference(0);

        LauncherManifest parent = getParentManifest();
        LOGGER.debug("Adding component to parent manifest");
        parent.getComponents().add(manifest.getId());
        String fileName = LauncherApplication.config.MANIFEST_FILE_NAME;
        if(manifest.getType() == LauncherManifestType.MODS_COMPONENT) {
            fileName = files.getGameDetailsManifest().getComponents().get(0);
        } else if(manifest.getType() == LauncherManifestType.SAVES_COMPONENT) {
            fileName = files.getGameDetailsManifest().getComponents().get(1);
        }
        LauncherFile.of(parent.getDirectory(), fileName).write(parent);
    }

    protected LauncherManifest getParentManifest() throws IOException {
        switch(manifest.getType()) {
            case SAVES_COMPONENT -> {
                return files.getSavesManifest();
            }
            case MODS_COMPONENT -> {
                return files.getModsManifest();
            }
            case RESOURCEPACKS_COMPONENT -> {
                return files.getResourcepackManifest();
            }
            case OPTIONS_COMPONENT -> {
                return files.getOptionsManifest();
            }
            case INSTANCE_COMPONENT ->  {
                return files.getInstanceManifest();
            }
            default -> throw new IOException("Invalid component type");
        }
    }

    protected void uploadAll(ComponentData data) throws IOException {
        SyncService service = new SyncService();
        uploadDirectory(data.getHashTree(), SyncService.convertType(manifest.getType()), manifest.getId(), manifest.getDirectory(), "", 0, data.getFileAmount(), service);
    }

    protected Integer uploadDirectory(List<ComponentData.HashEntry> entries, String type, String id, String basePath, String filePath, Integer currentAmount, Integer totalAmount, SyncService service) throws IOException {
        for(ComponentData.HashEntry child : entries) {
            if(child.getChildren() == null) {
                LOGGER.debug("Uploading file: " + child.getPath());
                currentAmount++;
                setStatus(new SyncStatus(SyncStep.UPLOADING, new DownloadStatus(currentAmount, totalAmount, child.getPath(), false)));
                LauncherFile file = LauncherFile.of(basePath, filePath, child.getPath());
                byte[] content = file.read();
                service.uploadFile(type, id, file.getPath(), content);
            } else {
                currentAmount = uploadDirectory(child.getChildren(), type, id, basePath, LauncherFile.of(filePath, child.getPath()).getPath(), currentAmount, totalAmount, service);
            }
        }
        return currentAmount;
    }

    protected void uploadDifference(List<String> difference) throws IOException {
        if(difference.isEmpty()) {
            LOGGER.debug("No difference found");
            return;
        }
        setStatus(new SyncStatus(SyncStep.UPLOADING, null));
        SyncService syncService = new SyncService();
        int index = 0;
        for(String path : difference) {
            index++;
            setStatus(new SyncStatus(SyncStep.UPLOADING, new DownloadStatus(index, difference.size(), path, false)));
            LOGGER.debug("Difference: " + path);
            LauncherFile file = LauncherFile.of(manifest.getDirectory(), path);
            byte[] content;
            if(file.isFile()) {
                content = file.read();
            } else {
                content = new byte[]{};
            }
            syncService.uploadFile(SyncService.convertType(manifest.getType()), manifest.getId(), path, content);
        }
    }

    protected int completeUpload(ComponentData newData) throws IOException {
        int version = new SyncService().complete(SyncService.convertType(manifest.getType()), manifest.getId());
        newData.setVersion(version);
        getSyncFile().write(newData);
        return version;
    }

    protected List<String> compareHashes(ComponentData.HashEntry oldEntry, ComponentData.HashEntry newEntry, LauncherFile path) {
        ArrayList<String> difference = new ArrayList<>();
        if(oldEntry.getChildren() == null && newEntry.getChildren() == null) {
            if(!oldEntry.getHash().equals(newEntry.getHash())) {
                LOGGER.debug("Adding changed file: " + path);
                difference.add(path.getPath());
            }
        } else if(oldEntry.getChildren() != null && newEntry.getChildren() != null) {
            int j = 0;
            for(int i = 0; i < oldEntry.getChildren().size(); i++) {
                boolean found = false;
                for(int k = j; k < newEntry.getChildren().size(); k++) {
                    if(oldEntry.getChildren().get(i).getPath().equals(newEntry.getChildren().get(k).getPath())) {
                        found = true;
                        for(int l = j; l < k; l++) {
                            LOGGER.debug("Adding added file: " + path + "/" + newEntry.getChildren().get(l).getPath());
                            difference.addAll(getAllChildren(newEntry.getChildren().get(l), LauncherFile.of(path, newEntry.getChildren().get(l).getPath())));
                        }
                        difference.addAll(compareHashes(oldEntry.getChildren().get(i), newEntry.getChildren().get(k), LauncherFile.of(path, oldEntry.getChildren().get(i).getPath())));
                        j = k+1;
                        break;
                    }
                }
                if(!found) {
                    LOGGER.debug("Adding deleted file: " + oldEntry.getChildren().get(i).getPath());
                    difference.addAll(getAllChildren(oldEntry.getChildren().get(i), LauncherFile.of(path,  oldEntry.getChildren().get(i).getPath())));
                }
            }
        } else if(oldEntry.getChildren() == null) {
            LOGGER.debug("Adding file that became folder: " + newEntry.getPath());
            difference.add(oldEntry.getPath());
            difference.addAll(getAllChildren(newEntry, path));
        } else {
            LOGGER.debug("Adding folder that became file: " + newEntry.getPath());
            difference.addAll(getAllChildren(oldEntry, path));
            difference.add(newEntry.getPath());
        }
        return difference;
    }

    protected List<String> getAllChildren(ComponentData.HashEntry entry, LauncherFile path) {
        ArrayList<String> children = new ArrayList<>();
        if(entry.getChildren() == null) {
            children.add(path.getPath());
        } else {
            for(ComponentData.HashEntry child : entry.getChildren()) {
                children.addAll(getAllChildren(child, LauncherFile.of(path, child.getPath())));
            }
        }
        return children;
    }

    protected ComponentData getCurrentComponentData() throws IOException {
        LauncherFile syncFile = LauncherFile.of(manifest.getDirectory(), LauncherApplication.config.SYNC_FILENAME);
        if(!syncFile.exists()) {
            throw new IOException("Sync file not found");
        }
        String componentData = syncFile.readString();
        try {
            return ComponentData.fromJson(componentData);
        } catch (SerializationException e) {
            throw new IOException("Failed to parse sync file", e);
        }
    }

    protected void downloadDifference(int currentVersion) throws IOException {
        setStatus(new SyncStatus(SyncStep.COLLECTING, null));
        GetResponse response = new SyncService().get(SyncService.convertType(manifest.getType()), manifest.getId(), currentVersion);
        if(response.getVersion() == currentVersion) {
            LOGGER.debug("Component is up to date");
            setStatus(new SyncStatus(SyncStep.FINISHED, null));
            return;
        }

        LOGGER.debug("Downloading component: version=" + currentVersion + " -> " + response.getVersion());
        setStatus(new SyncStatus(SyncStep.DOWNLOADING, null));

        downloadFiles(response.getDifference());

        LOGGER.debug("Updating sync file");
        setStatus(new SyncStatus(SyncStep.COLLECTING, null));
        updateSyncFile(response.getVersion());
        LOGGER.debug("Update complete");
        setStatus(new SyncStatus(SyncStep.FINISHED, null));
    }

    protected void downloadFiles(List<String> difference) throws IOException {
        SyncService service = new SyncService();
        String type = SyncService.convertType(manifest.getType());
        String basePath = manifest.getDirectory();
        int amount = 0;
        for(String path : difference) {
            try {
                path = UrlString.decoded(path).get();
            } catch (FormatString.FormatException e) {
                LOGGER.warn("Unable to decode filepath: " + path + ", this may be due to no url encoding being used, continuing with possibly encoded path, error: " + e);
            }
            LOGGER.debug("Downloading file: " + path);
            setStatus(new SyncStatus(SyncStep.DOWNLOADING, new DownloadStatus(++amount, difference.size(), path, false)));
            byte[] content = service.downloadFile(type, manifest.getId(), path);
            LauncherFile file = LauncherFile.of(basePath, path);
            if(content.length == 0) {
                if(file.isFile()) {
                    LOGGER.debug("Deleting file or dir: " + path);
                    file.remove();
                }
            } else {
                if(!file.isFile())  {
                    if(file.isDirectory()) {
                        file.remove();
                    }
                    file.createFile();
                }
                file.write(content);
            }
        }
    }

    protected ComponentData updateSyncFile(int version) throws IOException {
        ComponentData data = calculateComponentData(version);
        getSyncFile().write(data);
        return data;
    }

    protected LauncherFile getSyncFile() {
        return LauncherFile.of(manifest.getDirectory(), LauncherApplication.config.SYNC_FILENAME);
    }

    protected ComponentData calculateComponentData(int version) throws IOException {
        LOGGER.debug("Collecting component data for component: " + manifest.getId());
        long startTime = System.currentTimeMillis();
        File componentDir = new File(manifest.getDirectory());
        resetCount();
        List<ComponentData.HashEntry> result = hashDirectoryContents(componentDir);
        LOGGER.debug("Component data collected in " + (System.currentTimeMillis() - startTime) + "ms");
        return new ComponentData(version, resetCount(), result);
    }

    protected List<ComponentData.HashEntry> hashDirectoryContents(File dir) throws IOException {
        File[] children = dir.listFiles((dir1, name) -> !name.equals("data.sync") && !name.equals(".included_files_old"));
        if(children == null) return List.of();
        ArrayList<Pair<File, Integer>> files = new ArrayList<>();
        ArrayList<ComponentData.HashEntry> hashTree = new ArrayList<>();
        for(int i = 0; i < children.length; i++) {
            files.add(new Pair<>(children[i], i));
            hashTree.add(null);
        }
        ArrayList<IOException> exceptions = new ArrayList<>();
        files.parallelStream().forEach((file) -> {
            if(file.getKey().isDirectory()) {
                List<ComponentData.HashEntry> result;
                try {
                    result = hashDirectoryContents(file.getKey());
                } catch (IOException e) {
                    exceptions.add(e);
                    return;
                }
                hashTree.set(file.getValue(), new ComponentData.HashEntry(file.getKey().getName(), result));
            } else if(file.getKey().isFile()) {
                addCount(1);
                try {
                    hashTree.set(file.getValue(), new ComponentData.HashEntry(file.getKey().getName(), LauncherFile.of(file.getKey()).hash()));
                } catch (IOException e) {
                    exceptions.add(e);
                }
            }
        });
        if(!exceptions.isEmpty()) {
            throw exceptions.get(0);
        }
        return hashTree;
    }

    private int count = 0;
    protected void addCount(int amount) {
        this.count += amount;
    }
    protected int resetCount() {
        int count = this.count;
        this.count = 0;
        return count;
    }

    public LauncherManifest getManifest() {
        return manifest;
    }

    public void setManifest(LauncherManifest manifest) {
        this.manifest = manifest;
    }

}

File: client/composeApp/src/commonMain/kotlin/net/treset/treelauncher/backend/sync/ManifestSynchronizer.kt
package net.treset.treelauncher.backend.sync

import io.github.oshai.kotlinlogging.KotlinLogging
import net.treset.mcdl.json.SerializationException
import net.treset.mcdl.util.DownloadStatus
import net.treset.treelauncher.backend.config.appConfig
import net.treset.treelauncher.backend.data.LauncherFiles
import net.treset.treelauncher.backend.data.manifest.ComponentManifest
import net.treset.treelauncher.backend.data.manifest.LauncherManifestType
import net.treset.treelauncher.backend.data.manifest.ParentManifest
import net.treset.treelauncher.backend.util.file.LauncherFile
import net.treset.treelauncher.backend.util.string.FormatString
import net.treset.treelauncher.backend.util.string.UrlString
import java.io.File
import java.io.IOException
import java.nio.file.Files

open class ManifestSynchronizer(var manifest: ComponentManifest, protected var files: LauncherFiles, callback: SyncCallback?) :
    FileSynchronizer(callback) {
    @Throws(IOException::class)
    override fun upload() {
        if (SyncService.isSyncing(manifest)) {
            uploadExisting()
        } else {
            uploadNew()
        }
    }

    @Throws(IOException::class)
    override fun download() {
        if (SyncService.isSyncing(manifest)) {
            downloadExisting()
        } else {
            downloadNew()
        }
    }

    @Throws(IOException::class)
    protected fun uploadExisting() {
        setStatus(SyncStatus(SyncStep.COLLECTING, null))
        val currentData = currentComponentData
        val newData = calculateComponentData(currentData.version)
        val difference = compareHashes(
            ComponentData.HashEntry("", currentData.hashTree),
            ComponentData.HashEntry("", newData.hashTree),
            LauncherFile.of("./")
        )
        uploadDifference(difference)
        completeUpload(newData)
        setStatus(SyncStatus(SyncStep.FINISHED, null))
    }

    @Throws(IOException::class)
    protected fun uploadNew() {
        setStatus(SyncStatus(SyncStep.STARTING, null))
        LOGGER.debug { "Sync file not found, creating new sync file" }
        setStatus(SyncStatus(SyncStep.COLLECTING, null))
        val data = calculateComponentData(0)
        setStatus(SyncStatus(SyncStep.UPLOADING, null))
        val service = SyncService()
        service.newComponent(SyncService.convertType(manifest.type), manifest.id)
        uploadAll(data)
        LOGGER.debug { "Completing upload..." }
        val version = completeUpload(data)
        LOGGER.debug { "Upload complete: version=$version" }
        setStatus(SyncStatus(SyncStep.FINISHED, null))
    }

    @Throws(IOException::class)
    protected fun downloadExisting() {
        val data = currentComponentData
        downloadDifference(data.version)
    }

    @Throws(IOException::class)
    protected open fun downloadNew() {
        val dir = File(manifest.directory)
        if (!dir.exists()) {
            Files.createDirectories(dir.toPath())
        }
        downloadDifference(0)
        val parent: ParentManifest = parentManifest
        LOGGER.debug { "Adding component to parent manifest" }
        parent.components.add(manifest.id)
        var fileName: String = appConfig().manifestFileName
        LauncherFile.of(parent.directory, fileName).write(parent)
    }

    @get:Throws(IOException::class)
    protected val parentManifest: ParentManifest
        get() = when (manifest.type) {
            LauncherManifestType.SAVES_COMPONENT -> {
                files.savesManifest
            }

            LauncherManifestType.MODS_COMPONENT -> {
                files.modsManifest
            }

            LauncherManifestType.RESOURCEPACKS_COMPONENT -> {
                files.resourcepackManifest
            }

            LauncherManifestType.OPTIONS_COMPONENT -> {
                files.optionsManifest
            }

            LauncherManifestType.INSTANCE_COMPONENT -> {
                files.instanceManifest
            }

            else -> throw IOException("Invalid component type")
        }

    @Throws(IOException::class)
    protected open fun uploadAll(data: ComponentData) {
        val service = SyncService()
        uploadDirectory(
            data.hashTree,
            SyncService.convertType(manifest.type),
            manifest.id,
            manifest.directory,
            "",
            0,
            data.fileAmount,
            service
        )
    }

    @Throws(IOException::class)
    protected fun uploadDirectory(
        entries: List<ComponentData.HashEntry>,
        type: String,
        id: String,
        basePath: String,
        filePath: String,
        currentAmount: Int,
        totalAmount: Int,
        service: SyncService
    ): Int {
        var current = currentAmount
        for (child in entries) {
            child.children?.let {
                current = uploadDirectory(
                    it,
                    type,
                    id,
                    basePath,
                    LauncherFile.of(filePath, child.path).path,
                    current,
                    totalAmount,
                    service
                )
            }?: {
                LOGGER.debug { "Uploading file: ${child.path}" }
                current++
                setStatus(SyncStatus(SyncStep.UPLOADING, DownloadStatus(current, totalAmount, child.path)))
                val file: LauncherFile = LauncherFile.of(basePath, filePath, child.path)
                val content: ByteArray = file.read()
                service.uploadFile(type, id, file.path, content)
            }
        }
        return current
    }

    @Throws(IOException::class)
    protected open fun uploadDifference(difference: List<String>) {
        if (difference.isEmpty()) {
            LOGGER.debug { "No difference found" }
            return
        }
        setStatus(SyncStatus(SyncStep.UPLOADING, null))
        val syncService = SyncService()
        var index = 0
        for (path in difference) {
            index++
            setStatus(SyncStatus(SyncStep.UPLOADING, DownloadStatus(index, difference.size, path)))
            LOGGER.debug { "Difference: $path" }
            val file: LauncherFile = LauncherFile.of(manifest.directory, path)
            val content: ByteArray = if (file.isFile()) {
                file.read()
            } else {
                byteArrayOf()
            }
            syncService.uploadFile(SyncService.convertType(manifest.type), manifest.id, path, content)
        }
    }

    @Throws(IOException::class)
    protected fun completeUpload(newData: ComponentData): Int {
        val version = SyncService().complete(SyncService.convertType(manifest.type), manifest.id)
        newData.version = version
        syncFile.write(newData)
        return version
    }

    protected fun compareHashes(
        oldEntry: ComponentData.HashEntry,
        newEntry: ComponentData.HashEntry,
        path: LauncherFile
    ): List<String> {
        val difference = ArrayList<String>()
        if (oldEntry.children == null && newEntry.children == null) {
            if (oldEntry.hash != newEntry.hash) {
                LOGGER.debug { "Adding changed file: $path" }
                difference.add(path.path)
            }
        } else if (oldEntry.children != null && newEntry.children != null) {
            var j = 0
            for (i in oldEntry.children!!.indices) {
                var found = false
                for (k in j until newEntry.children!!.size) {
                    if (oldEntry.children!![i].path == newEntry.children!![k].path) {
                        found = true
                        for (l in j until k) {
                            LOGGER.debug { "Adding added file: $path/${newEntry.children!![l].path}" }
                            difference.addAll(
                                getAllChildren(
                                    newEntry.children!![l],
                                    LauncherFile.of(path, newEntry.children!![l].path)
                                )
                            )
                        }
                        difference.addAll(
                            compareHashes(
                                oldEntry.children!![i],
                                newEntry.children!![k],
                                LauncherFile.of(path, oldEntry.children!![i].path)
                            )
                        )
                        j = k + 1
                        break
                    }
                }
                if (!found) {
                    LOGGER.debug { "Adding deleted file: ${oldEntry.children!![i].path}" }
                    difference.addAll(
                        getAllChildren(
                            oldEntry.children!![i],
                            LauncherFile.of(path, oldEntry.children!![i].path)
                        )
                    )
                }
            }
        } else if (oldEntry.children == null) {
            LOGGER.debug { "Adding file that became folder: ${newEntry.path}" }
            difference.add(oldEntry.path)
            difference.addAll(getAllChildren(newEntry, path))
        } else {
            LOGGER.debug { "Adding folder that became file: ${newEntry.path}" }
            difference.addAll(getAllChildren(oldEntry, path))
            difference.add(newEntry.path)
        }
        return difference
    }

    private fun getAllChildren(entry: ComponentData.HashEntry, path: LauncherFile): List<String> {
        val children = ArrayList<String>()
        if (entry.children == null) {
            children.add(path.path)
        } else {
            for (child in entry.children!!) {
                children.addAll(getAllChildren(child, LauncherFile.of(path, child.path)))
            }
        }
        return children
    }

    @get:Throws(IOException::class)
    protected val currentComponentData: ComponentData
        get() {
            val syncFile: LauncherFile = LauncherFile.of(manifest.directory, appConfig().syncFileName)
            if (!syncFile.exists()) {
                throw IOException("Sync file not found")
            }
            val componentData: String = syncFile.readString()
            return try {
                ComponentData.fromJson(componentData)
            } catch (e: SerializationException) {
                throw IOException("Failed to parse sync file", e)
            }
        }

    @Throws(IOException::class)
    protected fun downloadDifference(currentVersion: Int) {
        setStatus(SyncStatus(SyncStep.COLLECTING, null))
        val response = SyncService()[SyncService.convertType(manifest.type), manifest.id, currentVersion]
        if (response.version == currentVersion) {
            LOGGER.debug { "Component is up to date" }
            setStatus(SyncStatus(SyncStep.FINISHED, null))
            return
        }
        LOGGER.debug { "Downloading component: version=$currentVersion -> $response.version" }
        setStatus(SyncStatus(SyncStep.DOWNLOADING, null))
        downloadFiles(response.difference)
        LOGGER.debug { "Updating sync file" }
        setStatus(SyncStatus(SyncStep.COLLECTING, null))
        updateSyncFile(response.version)
        LOGGER.debug { "Update complete" }
        setStatus(SyncStatus(SyncStep.FINISHED, null))
    }

    @Throws(IOException::class)
    protected open fun downloadFiles(difference: Array<String>) {
        val service = SyncService()
        val type = SyncService.convertType(manifest.type)
        val basePath: String = manifest.directory
        for ((amount, rawPath) in difference.withIndex()) {
            val path: String = try {
                UrlString.decoded(rawPath).get()
            } catch (e: FormatString.FormatException) {
                LOGGER.warn { "Unable to decode filepath: $rawPath, this may be due to no url encoding being used, continuing with possibly encoded path, error: $e" }
                rawPath
            }
            LOGGER.debug { "Downloading file: $path" }
            setStatus(SyncStatus(SyncStep.DOWNLOADING, DownloadStatus(amount + 1, difference.size, path)))
            val content = service.downloadFile(type, manifest.id, path)
            val file: LauncherFile = LauncherFile.of(basePath, path)
            if (content.isEmpty()) {
                if (file.isFile()) {
                    LOGGER.debug { "Deleting file or dir: $path" }
                    file.remove()
                }
            } else {
                if (!file.isFile()) {
                    if (file.isDirectory()) {
                        file.remove()
                    }
                    file.createFile()
                }
                file.write(content)
            }
        }
    }

    @Throws(IOException::class)
    protected fun updateSyncFile(version: Int): ComponentData {
        val data = calculateComponentData(version)
        syncFile.write(data)
        return data
    }

    private val syncFile: LauncherFile
        get() = LauncherFile.of(manifest.directory, appConfig().syncFileName)

    @Throws(IOException::class)
    protected fun calculateComponentData(version: Int): ComponentData {
        LOGGER.debug { "Collecting component data for component: ${manifest.id}" }
        val startTime = System.currentTimeMillis()
        val componentDir = File(manifest.directory)
        resetCount()
        val result = hashDirectoryContents(componentDir)
        LOGGER.debug { "Component data collected in ${System.currentTimeMillis() - startTime}ms" }
        return ComponentData(version, resetCount(), result)
    }

    @Throws(IOException::class)
    protected fun hashDirectoryContents(dir: File): List<ComponentData.HashEntry> {
        val children = dir.listFiles { _: File?, name: String -> name != "data.sync" && name != ".included_files_old" } ?: return listOf()
        val files: MutableList<Pair<File, Int>> = mutableListOf()
        val hashTree: MutableList<ComponentData.HashEntry> = mutableListOf()
        for (i in children.indices) {
            files.add(Pair(children[i], i))
        }
        val exceptions: MutableList<IOException> = mutableListOf()
        files.parallelStream().forEach { file: Pair<File, Int> ->
            if (file.first.isDirectory()) {
                val result: List<ComponentData.HashEntry> = try {
                    hashDirectoryContents(file.first)
                } catch (e: IOException) {
                    exceptions.add(e)
                    return@forEach
                }
                hashTree[file.second] = ComponentData.HashEntry(file.first.getName(), result)
            } else if (file.first.isFile()) {
                addCount(1)
                try {
                    hashTree[file.second] = ComponentData.HashEntry(file.first.getName(), LauncherFile.of(file.first).hash())
                } catch (e: IOException) {
                    exceptions.add(e)
                }
            }
        }
        if (exceptions.isNotEmpty()) {
            throw exceptions[0]
        }
        return hashTree
    }

    private var count = 0

    protected fun addCount(amount: Int) {
        count += amount
    }

    protected fun resetCount(): Int {
        val count = count
        this.count = 0
        return count
    }

    companion object {
        private val LOGGER = KotlinLogging.logger {}
    }
}
